In [1]:
import pandas as pd
import numpy as np

import warnings
warnings.filterwarnings('ignore')
import itertools

from detailed_details import detailed_details

from IPython.core.interactiveshell import InteractiveShell
InteractiveShell.ast_node_interactivity = "all"

In [2]:
unique_SSN = []

# open file and read the content in a list
with open(r'list_files//unique_SSN.txt', 'r') as fp:
    for line in fp:
        # remove linebreak from a current name
        # linebreak is the last character of each line
        x = line[:-1]

        # add current item to the list
        unique_SSN.append(int(x))

In [17]:
df = pd.read_csv('credit_train.csv')

In [18]:
df.columns

Index(['ID', 'Customer_ID', 'Month', 'Name', 'Age', 'SSN', 'Occupation',
       'Annual_Income', 'Monthly_Inhand_Salary', 'Num_Bank_Accounts',
       'Num_Credit_Card', 'Interest_Rate', 'Num_of_Loan', 'Type_of_Loan',
       'Delay_from_due_date', 'Num_of_Delayed_Payment', 'Changed_Credit_Limit',
       'Num_Credit_Inquiries', 'Credit_Mix', 'Outstanding_Debt',
       'Credit_Utilization_Ratio', 'Credit_History_Age',
       'Payment_of_Min_Amount', 'Total_EMI_per_month',
       'Amount_invested_monthly', 'Payment_Behaviour', 'Monthly_Balance',
       'Credit_Score'],
      dtype='object')

# Preprocessing

In [19]:
df.head()

ID Customer_ID     Month           Name   Age          SSN Occupation  \
0  0x1602   CUS_0xd40   January  Aaron Maashoh    23  821-00-0265  Scientist   
1  0x1603   CUS_0xd40  February  Aaron Maashoh    23  821-00-0265  Scientist   
2  0x1604   CUS_0xd40     March  Aaron Maashoh  -500  821-00-0265  Scientist   
3  0x1605   CUS_0xd40     April  Aaron Maashoh    23  821-00-0265  Scientist   
4  0x1606   CUS_0xd40       May  Aaron Maashoh    23  821-00-0265  Scientist   

  Annual_Income  Monthly_Inhand_Salary  Num_Bank_Accounts  ...  Credit_Mix  \
0      19114.12            1824.843333                  3  ...           _   
1      19114.12                    NaN                  3  ...        Good   
2      19114.12                    NaN                  3  ...        Good   
3      19114.12                    NaN                  3  ...        Good   
4      19114.12            1824.843333                  3  ...        Good   

   Outstanding_Debt Credit_Utilization_Ratio     Credit_History_Age  \
0            809.98                26.822620  22 Years and 1 Months   
1            809.98                31.944960                    NaN   
2            809.98                28.609352  22 Years and 3 Months   
3            809.98                31.377862  22 Years and 4 Months   
4            809.98                24.797347  22 Years and 5 Months   

   Payment_of_Min_Amount Total_EMI_per_month Amount_invested_monthly  \
0                     No           49.574949       80.41529543900253   
1                     No           49.574949      118.28022162236736   
2                     No           49.574949         81.699521264648   
3                     No           49.574949       199.4580743910713   
4                     No           49.574949      41.420153086217326   

                  Payment_Behaviour     Monthly_Balance Credit_Score  
0   High_spent_Small_value_payments  312.49408867943663         Good  
1    Low_spent_Large_value_payments  284.62916249607184         Good  
2   Low_spent_Medium_value_payments   331.2098628537912         Good  
3    Low_spent_Small_value_payments  223.45130972736786         Good  
4  High_spent_Medium_value_payments  341.48923103222177         Good  

[5 rows x 28 columns]

### PreProcessing ID and Customer_ID

In [20]:
df_id = df.copy()

In [21]:
df_id['ID'].nunique()

100000

##### As ID is unique for all values. We can drop them.

In [22]:
df_id.drop(columns = 'ID', axis = 1, inplace = True)

In [23]:
df_id.head()

Customer_ID     Month           Name   Age          SSN Occupation  \
0   CUS_0xd40   January  Aaron Maashoh    23  821-00-0265  Scientist   
1   CUS_0xd40  February  Aaron Maashoh    23  821-00-0265  Scientist   
2   CUS_0xd40     March  Aaron Maashoh  -500  821-00-0265  Scientist   
3   CUS_0xd40     April  Aaron Maashoh    23  821-00-0265  Scientist   
4   CUS_0xd40       May  Aaron Maashoh    23  821-00-0265  Scientist   

  Annual_Income  Monthly_Inhand_Salary  Num_Bank_Accounts  Num_Credit_Card  \
0      19114.12            1824.843333                  3                4   
1      19114.12                    NaN                  3                4   
2      19114.12                    NaN                  3                4   
3      19114.12                    NaN                  3                4   
4      19114.12            1824.843333                  3                4   

   ...  Credit_Mix Outstanding_Debt Credit_Utilization_Ratio  \
0  ...           _           809.98                26.822620   
1  ...        Good           809.98                31.944960   
2  ...        Good           809.98                28.609352   
3  ...        Good           809.98                31.377862   
4  ...        Good           809.98                24.797347   

      Credit_History_Age Payment_of_Min_Amount Total_EMI_per_month  \
0  22 Years and 1 Months                    No           49.574949   
1                    NaN                    No           49.574949   
2  22 Years and 3 Months                    No           49.574949   
3  22 Years and 4 Months                    No           49.574949   
4  22 Years and 5 Months                    No           49.574949   

   Amount_invested_monthly                 Payment_Behaviour  \
0        80.41529543900253   High_spent_Small_value_payments   
1       118.28022162236736    Low_spent_Large_value_payments   
2          81.699521264648   Low_spent_Medium_value_payments   
3        199.4580743910713    Low_spent_Small_value_payments   
4       41.420153086217326  High_spent_Medium_value_payments   

      Monthly_Balance  Credit_Score  
0  312.49408867943663          Good  
1  284.62916249607184          Good  
2   331.2098628537912          Good  
3  223.45130972736786          Good  
4  341.48923103222177          Good  

[5 rows x 27 columns]

Customer_ID, Name, SSN all seems to be same

In [24]:
df_id['Customer_ID'].nunique()
df_id["SSN"].nunique()

12500

12501

Check How many unique SSN each customer_ID has

In [25]:
df_id.groupby(['Customer_ID', 'SSN'])['SSN'].unique()

Customer_ID  SSN        
CUS_0x1000   913-74-1218    [913-74-1218]
CUS_0x1009   #F%$D@*&8        [#F%$D@*&8]
             063-67-6938    [063-67-6938]
CUS_0x100b   #F%$D@*&8        [#F%$D@*&8]
             238-62-0395    [238-62-0395]
                                ...      
CUS_0xff3    726-35-5322    [726-35-5322]
CUS_0xff4    655-05-7666    [655-05-7666]
CUS_0xff6    541-92-8371    [541-92-8371]
CUS_0xffc    226-86-7294    [226-86-7294]
CUS_0xffd    832-88-8320    [832-88-8320]
Name: SSN, Length: 17115, dtype: object

In [26]:
unique_values = list(df_id[df_id['Customer_ID'] == 'CUS_0xd40']['SSN'].unique())
unique_values

['821-00-0265', '#F%$D@*&8']

Each Customer_ID seems to have a unique SSN except for some noisy data (#F%$D@*&8)

### PreProcessing SSN

In [27]:
df_ssn = df_id.copy()

In [28]:
(df_ssn["SSN"] == '#F%$D@*&8').sum()

5572

In [29]:
corresponding_unique_ssn_values = {}

In [30]:
unique_customer_id = list(df_ssn['Customer_ID'].unique())

In [31]:
list(df_ssn[df_ssn['Customer_ID'] == 'CUS_0xd40']['SSN'].unique())

['821-00-0265', '#F%$D@*&8']

Replace SSN according to corresponding Customer_ID

In [32]:
for id in unique_customer_id:
    corresponding_unique_ssn_values[id] = list(df_ssn[df_ssn['Customer_ID'] == id]['SSN'].unique())

In [33]:
dict(itertools.islice(corresponding_unique_ssn_values.items(), 5))

{'CUS_0xd40': ['821-00-0265', '#F%$D@*&8'],
 'CUS_0x21b1': ['004-07-5839'],
 'CUS_0x2dbc': ['486-85-3974'],
 'CUS_0xb891': ['072-31-6145', '#F%$D@*&8'],
 'CUS_0x1cdb': ['615-06-7821']}

In [34]:
(df_ssn['SSN'] == '#F%$D@*&8').sum()

count = 0
for i in range(len(df_ssn)):
    if(df_ssn['SSN'][i].startswith('#')):
        count += 1 
count

5572

5572

In [35]:
[val for val in corresponding_unique_ssn_values[df_ssn['Customer_ID'][0]] if not val.startswith('#')]
[val for val in corresponding_unique_ssn_values['CUS_0xd40'] if not val.startswith('#')]

['821-00-0265']

['821-00-0265']

We can thus deduce that all noisy data starts with "#"

Replacing values of noisy data

In [36]:
for index in range(len(df_ssn)):
    if df_ssn['SSN'][index].startswith('#'):
        df_ssn['SSN'][index] = [val for val in corresponding_unique_ssn_values[df_ssn['Customer_ID'][index]] if not val.startswith('#')][0]

In [ ]:
df_ssn.head()

Customer_ID     Month           Name   Age          SSN Occupation  \
0   CUS_0xd40   January  Aaron Maashoh    23  821-00-0265  Scientist   
1   CUS_0xd40  February  Aaron Maashoh    23  821-00-0265  Scientist   
2   CUS_0xd40     March  Aaron Maashoh  -500  821-00-0265  Scientist   
3   CUS_0xd40     April  Aaron Maashoh    23  821-00-0265  Scientist   
4   CUS_0xd40       May  Aaron Maashoh    23  821-00-0265  Scientist   

  Annual_Income  Monthly_Inhand_Salary  Num_Bank_Accounts  Num_Credit_Card  \
0      19114.12            1824.843333                  3                4   
1      19114.12                    NaN                  3                4   
2      19114.12                    NaN                  3                4   
3      19114.12                    NaN                  3                4   
4      19114.12            1824.843333                  3                4   

   ...  Credit_Mix Outstanding_Debt Credit_Utilization_Ratio  \
0  ...           _           809.98                26.822620   
1  ...        Good           809.98                31.944960   
2  ...        Good           809.98                28.609352   
3  ...        Good           809.98                31.377862   
4  ...        Good           809.98                24.797347   

      Credit_History_Age Payment_of_Min_Amount Total_EMI_per_month  \
0  22 Years and 1 Months                    No           49.574949   
1                    NaN                    No           49.574949   
2  22 Years and 3 Months                    No           49.574949   
3  22 Years and 4 Months                    No           49.574949   
4  22 Years and 5 Months                    No           49.574949   

   Amount_invested_monthly                 Payment_Behaviour  \
0        80.41529543900253   High_spent_Small_value_payments   
1       118.28022162236736    Low_spent_Large_value_payments   
2          81.699521264648   Low_spent_Medium_value_payments   
3        199.4580743910713    Low_spent_Small_value_payments   
4       41.420153086217326  High_spent_Medium_value_payments   

      Monthly_Balance  Credit_Score  
0  312.49408867943663          Good  
1  284.62916249607184          Good  
2   331.2098628537912          Good  
3  223.45130972736786          Good  
4  341.48923103222177          Good  

[5 rows x 27 columns]

In [ ]:
for id in unique_customer_id:
    corresponding_unique_ssn_values[id] = list(df_ssn[df_ssn['Customer_ID'] == id]['SSN'].unique())
dict(itertools.islice(corresponding_unique_ssn_values.items(), 5))

{'CUS_0xd40': ['821-00-0265'],
 'CUS_0x21b1': ['004-07-5839'],
 'CUS_0x2dbc': ['486-85-3974'],
 'CUS_0xb891': ['072-31-6145'],
 'CUS_0x1cdb': ['615-06-7821']}

making SSN more Readable by replacing - with empty value and making it a int value

In [ ]:
for index in range(len(df_ssn)):
    df_ssn['SSN'][index] = int (df_ssn['SSN'][index].replace('-',''))
    

In [ ]:
df_ssn['SSN'].nunique()

12500

Now dropping column Customer_ID because each row can now be identified by ssn

In [ ]:
df_ssn.drop(columns = ['Customer_ID'],axis = 1, inplace = True)

In [ ]:
df_ssn.to_csv('checkpoints//df_ssn.csv',index = False )

### PreProcessing Month

In [2]:
df_month = pd.read_csv('checkpoints//df_ssn.csv')

In [3]:
df_month['Month'].value_counts()

January     12500
February    12500
March       12500
April       12500
May         12500
June        12500
July        12500
August      12500
Name: Month, dtype: int64

Leaving months as it is for now

### PreProcessing Name

In [3]:
df_name = df_month.copy()

In [4]:
df_name['Name'].nunique()

10139

name maynot be that relevant so dropping

In [5]:
df_name.drop(columns = ['Name'], axis = 1, inplace = True)

In [6]:
df_name.head()

Month   Age        SSN Occupation Annual_Income  Monthly_Inhand_Salary  \
0   January    23  821000265  Scientist      19114.12            1824.843333   
1  February    23  821000265  Scientist      19114.12                    NaN   
2     March  -500  821000265  Scientist      19114.12                    NaN   
3     April    23  821000265  Scientist      19114.12                    NaN   
4       May    23  821000265  Scientist      19114.12            1824.843333   

   Num_Bank_Accounts  Num_Credit_Card  Interest_Rate Num_of_Loan  ...  \
0                  3                4              3           4  ...   
1                  3                4              3           4  ...   
2                  3                4              3           4  ...   
3                  3                4              3           4  ...   
4                  3                4              3           4  ...   

  Credit_Mix  Outstanding_Debt Credit_Utilization_Ratio  \
0          _            809.98                26.822620   
1       Good            809.98                31.944960   
2       Good            809.98                28.609352   
3       Good            809.98                31.377862   
4       Good            809.98                24.797347   

      Credit_History_Age  Payment_of_Min_Amount Total_EMI_per_month  \
0  22 Years and 1 Months                     No           49.574949   
1                    NaN                     No           49.574949   
2  22 Years and 3 Months                     No           49.574949   
3  22 Years and 4 Months                     No           49.574949   
4  22 Years and 5 Months                     No           49.574949   

  Amount_invested_monthly                 Payment_Behaviour  \
0       80.41529543900253   High_spent_Small_value_payments   
1      118.28022162236736    Low_spent_Large_value_payments   
2         81.699521264648   Low_spent_Medium_value_payments   
3       199.4580743910713    Low_spent_Small_value_payments   
4      41.420153086217326  High_spent_Medium_value_payments   

      Monthly_Balance Credit_Score  
0  312.49408867943663         Good  
1  284.62916249607184         Good  
2   331.2098628537912         Good  
3  223.45130972736786         Good  
4  341.48923103222177         Good  

[5 rows x 25 columns]

### PreProcessing Age

In [8]:
df_age = df_name.copy()

In [9]:
df_age['Age'].isna().sum()

0

Count possible noisy data in Age (age upper limit = 120 and lower limit is 0)

In [10]:
underscore_age = 0
underscore_age_list = []
noisy_age = 0
noisy_age_list = []
for index in range(len(df_age)):
    if '_' in  df_age['Age'][index]:
        underscore_age += 1
        underscore_age_list.append(df_age['Age'][index])
    try:
        if int(df_age['Age'][index]) > 120 or int(df_age['Age'][index] <= 0):
            noisy_age +=1
            noisy_age_list.append(df_age['Age'][index])
            
    except:
        pass
print("Underscore_age Count : ",underscore_age,"|| Unique Underscore_age count :",len(set(underscore_age_list)), " || Example : ",underscore_age_list[0:5])
print("Noisy_age Count : ",noisy_age,"|| Unique Noisy_age count :",len(set(noisy_age_list))," || Example : ",noisy_age_list[0:5])



Underscore_age Count :  4939 || Unique Underscore_age count : 127  || Example :  ['28_', '34_', '30_', '24_', '33_']
Noisy_age Count :  1799 || Unique Noisy_age count : 1608  || Example :  ['7580', '181', '995', '5079', '6409']


In [11]:
df_age['Age'].nunique()

1788

Probable Unique age values <br>
1788 - 127 - 1608 =53

Ommiting the underscore ages

In [12]:
for index in range(len(df_age)):
    if '_' in df_age['Age'][index]:
        df_age['Age'][index] = df_age['Age'][index].replace('_','')

Ommiting noisy ages

In [9]:
corresponding_unique_age_values= {}
unique_SSN = list(df_age['SSN'].unique())

In [ ]:
df_age['Age'] = df_age['Age'].astype(int)

In [ ]:
for ssn in unique_SSN:
    corresponding_unique_age_values[ssn] = list(df_age[df_age['SSN'] == ssn]['Age'].unique())

In [ ]:
dict(itertools.islice(corresponding_unique_age_values.items(),10))

{821000265: [23, -500],
 4075839: [28],
 486853974: [34],
 72316145: [54, 55],
 615067821: [21],
 612708987: [31],
 411510676: [33, 34],
 500926408: [7580, 30],
 70191622: [23, 24],
 366681681: [44, 45]}

For now allowing multiple valid ages but replacing the noisy data

In [ ]:
[val for val in corresponding_unique_age_values[df_age['SSN'][455]] if ((int(val) >= 0) and (int(val) <= 120)) ][0]

25

In [ ]:
for index in range(len(df_age)):
    if df_age['Age'][index] > 120 or df_age['Age'][index] <= 0:
        df_age['Age'][index] = [val for val in corresponding_unique_age_values[df_age['SSN'][index]] if ((val > 0) and (val <= 120)) ][0]

Lets Check

In [ ]:
noisy_age = 0
noisy_age_list = []
for index in range(len(df_age)):
    if df_age['Age'][index] > 120 or df_age['Age'][index] <= 0:
        noisy_age +=1
        noisy_age_list.append(df_age['Age'][index])
print("Noisy_age Count : ",noisy_age,"|| Unique Noisy_age count :",len(set(noisy_age_list))," || Example : ",noisy_age_list[0:5])

Noisy_age Count :  0 || Unique Noisy_age count : 0  || Example :  []


In [ ]:
df_age['Age'].value_counts()

38     3076
28     3057
31     3043
26     3035
32     2968
36     2955
25     2952
35     2937
39     2932
27     2929
34     2922
44     2895
22     2876
19     2874
41     2863
20     2834
37     2824
29     2819
43     2814
21     2801
30     2799
24     2786
23     2725
45     2711
40     2687
42     2648
33     2624
18     2446
46     1663
15     1611
17     1551
16     1502
48     1427
49     1413
55     1396
53     1391
52     1391
54     1343
51     1319
50     1314
47     1250
14     1223
56      363
95        3
112       1
118       1
99        1
100       1
102       1
111       1
109       1
115       1
Name: Age, dtype: int64

In [ ]:
df_age.to_csv('checkpoints//df_age.csv',index = False)

### PreProcessing Occupation

In [ ]:
df_occupation = pd.read_csv('checkpoints//df_age.csv')

In [ ]:
df_occupation.head()

Month  Age        SSN Occupation Annual_Income  Monthly_Inhand_Salary  \
0   January   23  821000265  Scientist      19114.12            1824.843333   
1  February   23  821000265  Scientist      19114.12                    NaN   
2     March   23  821000265  Scientist      19114.12                    NaN   
3     April   23  821000265  Scientist      19114.12                    NaN   
4       May   23  821000265  Scientist      19114.12            1824.843333   

   Num_Bank_Accounts  Num_Credit_Card  Interest_Rate Num_of_Loan  ...  \
0                  3                4              3           4  ...   
1                  3                4              3           4  ...   
2                  3                4              3           4  ...   
3                  3                4              3           4  ...   
4                  3                4              3           4  ...   

  Credit_Mix  Outstanding_Debt Credit_Utilization_Ratio  \
0          _            809.98                26.822620   
1       Good            809.98                31.944960   
2       Good            809.98                28.609352   
3       Good            809.98                31.377862   
4       Good            809.98                24.797347   

      Credit_History_Age  Payment_of_Min_Amount Total_EMI_per_month  \
0  22 Years and 1 Months                     No           49.574949   
1                    NaN                     No           49.574949   
2  22 Years and 3 Months                     No           49.574949   
3  22 Years and 4 Months                     No           49.574949   
4  22 Years and 5 Months                     No           49.574949   

  Amount_invested_monthly                 Payment_Behaviour  \
0       80.41529543900253   High_spent_Small_value_payments   
1      118.28022162236736    Low_spent_Large_value_payments   
2         81.699521264648   Low_spent_Medium_value_payments   
3       199.4580743910713    Low_spent_Small_value_payments   
4      41.420153086217326  High_spent_Medium_value_payments   

      Monthly_Balance Credit_Score  
0  312.49408867943663         Good  
1  284.62916249607184         Good  
2   331.2098628537912         Good  
3  223.45130972736786         Good  
4  341.48923103222177         Good  

[5 rows x 25 columns]

In [ ]:
df_occupation["Occupation"].nunique()

16

In [ ]:
df_occupation["Occupation"].value_counts()

_______          7062
Lawyer           6575
Architect        6355
Engineer         6350
Scientist        6299
Mechanic         6291
Accountant       6271
Developer        6235
Media_Manager    6232
Teacher          6215
Entrepreneur     6174
Doctor           6087
Journalist       6085
Manager          5973
Musician         5911
Writer           5885
Name: Occupation, dtype: int64

_______ seems to be noisy datas

In [ ]:
corresponding_unique_occupation_values = {}

In [ ]:
for ssn in unique_SSN:
    corresponding_unique_occupation_values[ssn] = list(df_occupation[df_occupation['SSN'] == ssn]['Occupation'].unique())

In [ ]:
dict(itertools.islice(corresponding_unique_occupation_values.items(),10))

{821000265: ['Scientist'],
 4075839: ['_______', 'Teacher'],
 486853974: ['_______', 'Engineer'],
 72316145: ['Entrepreneur', '_______'],
 615067821: ['Developer'],
 612708987: ['Lawyer', '_______'],
 411510676: ['Lawyer'],
 500926408: ['Media_Manager', '_______'],
 70191622: ['Doctor'],
 366681681: ['Journalist']}

In [ ]:
[val for val in corresponding_unique_occupation_values[df_occupation['SSN'][1]] if not val.startswith('_') ][0]

'Scientist'

In [ ]:
for index in range(len(df_occupation)):
    if df_occupation['Occupation'][index].startswith('_'):
        df_occupation['Occupation'][index] = [val for val in corresponding_unique_occupation_values[df_occupation['SSN'][index]] if not val.startswith('_') ][0]

In [ ]:
df_occupation["Occupation"].value_counts()

Lawyer           7096
Engineer         6864
Architect        6824
Mechanic         6776
Scientist        6744
Accountant       6744
Developer        6720
Media_Manager    6720
Teacher          6672
Entrepreneur     6648
Doctor           6568
Journalist       6536
Manager          6432
Musician         6352
Writer           6304
Name: Occupation, dtype: int64

In [ ]:
for ssn in unique_SSN:
    corresponding_unique_occupation_values[ssn] = list(df_occupation[df_occupation['SSN'] == ssn]['Occupation'].unique())
dict(itertools.islice(corresponding_unique_occupation_values.items(),10))

{821000265: ['Scientist'],
 4075839: ['Teacher'],
 486853974: ['Engineer'],
 72316145: ['Entrepreneur'],
 615067821: ['Developer'],
 612708987: ['Lawyer'],
 411510676: ['Lawyer'],
 500926408: ['Media_Manager'],
 70191622: ['Doctor'],
 366681681: ['Journalist']}

### PreProcessing Annual Income

In [ ]:
df_annual_income = df_occupation.copy()

In [ ]:
df_annual_income["Annual_Income"].nunique()

18940

In [ ]:
underscore_annual_income = 0
underscore_annual_income_list = []
for index in range(len(df_annual_income)):
    if '_' in  df_annual_income['Annual_Income'][index]:
        underscore_annual_income += 1
        underscore_annual_income_list.append(df_annual_income['Annual_Income'][index])
print("Underscore_annual_income Count : ",underscore_annual_income,"|| Unique Underscore_annual_income count :",len(set(underscore_annual_income_list)), " || Example : ",underscore_annual_income_list[0:5])

Underscore_annual_income Count :  6980 || Unique Underscore_annual_income count : 5503  || Example :  ['34847.84_', '30689.89_', '35547.71_', '34081.38_', '114838.41_']


Ommitting the underscore annual income

In [ ]:
for index in range(len(df_annual_income)):
    if '_' in df_annual_income['Annual_Income'][index]:
        df_annual_income['Annual_Income'][index] = df_annual_income['Annual_Income'][index].replace('_','')

In [ ]:
underscore_annual_income = 0
underscore_annual_income_list = []
for index in range(len(df_annual_income)):
    if '_' in  df_annual_income['Annual_Income'][index]:
        underscore_annual_income += 1
        underscore_annual_income_list.append(df_annual_income['Annual_Income'][index])
print("Underscore_annual_income Count : ",underscore_annual_income,"|| Unique Underscore_annual_income count :",len(set(underscore_annual_income_list)), " || Example : ",underscore_annual_income_list[0:5])

Underscore_annual_income Count :  0 || Unique Underscore_annual_income count : 0  || Example :  []


In [ ]:
df_annual_income['Annual_Income'] = df_annual_income['Annual_Income'].astype(float)

In [ ]:
df_annual_income.to_csv('checkpoints//df_annual_income.csv',index = False)

### PreProcessing Monthly_Inhand_Salary

In [10]:
df_monthly_inhand_salary = pd.read_csv('checkpoints//df_annual_income.csv')

In [11]:
df_monthly_inhand_salary['Monthly_Inhand_Salary'].nunique()

13235

In [12]:
df_monthly_inhand_salary['Monthly_Inhand_Salary'].isna().sum()

15002

Checking if each customer have same monthly inhand salary

In [13]:
corresponding_unique_inhand_salary_values = {}

In [14]:
for ssn in unique_SSN:
    corresponding_unique_inhand_salary_values[ssn] = list(df_monthly_inhand_salary[df_monthly_inhand_salary['SSN'] == ssn]['Monthly_Inhand_Salary'].unique())

In [15]:
dict(itertools.islice(corresponding_unique_inhand_salary_values.items(),10))

{821000265: [1824.8433333333328, nan],
 4075839: [3037.986666666666, nan],
 486853974: [12187.22, nan],
 72316145: [2612.4908333333333],
 615067821: [2853.3091666666664, nan],
 612708987: [nan, 5988.705000000001],
 411510676: [11242.783333333333, 10469.207759388815, nan],
 500926408: [nan, 2611.115],
 70191622: [9843.8675, nan],
 366681681: [2825.233333333333, nan]}

In [26]:
[val for val in corresponding_unique_inhand_salary_values[df_monthly_inhand_salary['SSN'][0]] if not np.isnan(val) ][0]

1824.8433333333328

In [29]:
for index in range(len(df_monthly_inhand_salary)):
    if np.isnan(df_monthly_inhand_salary['Monthly_Inhand_Salary'][index]):
        df_monthly_inhand_salary['Monthly_Inhand_Salary'][index] = [val for val in corresponding_unique_inhand_salary_values[df_monthly_inhand_salary['SSN'][index]] if not np.isnan(val) ][0]

In [30]:
for ssn in unique_SSN:
    corresponding_unique_inhand_salary_values[ssn] = list(df_monthly_inhand_salary[df_monthly_inhand_salary['SSN'] == ssn]['Monthly_Inhand_Salary'].unique())
dict(itertools.islice(corresponding_unique_inhand_salary_values.items(),10))

{821000265: [1824.8433333333328],
 4075839: [3037.986666666666],
 486853974: [12187.22],
 72316145: [2612.4908333333333],
 615067821: [2853.3091666666664],
 612708987: [5988.705000000001],
 411510676: [11242.783333333333, 10469.207759388815],
 500926408: [2611.115],
 70191622: [9843.8675],
 366681681: [2825.233333333333]}

In [36]:
df_monthly_inhand_salary.to_csv('checkpoints//df_monthly_inhand_salary.csv',index = False)

In [97]:
df_monthly_inhand_salary['Monthly_Inhand_Salary'].isna().sum()

0

### PreProcessing Num_Bank_Accounts

In [109]:
df_num_bank_accounts = pd.read_csv('checkpoints//df_monthly_inhand_salary.csv')

In [110]:
df_num_bank_accounts['Num_Bank_Accounts'].nunique()

943

In [111]:
corresponding_unique_bank_accounts_values = {}

In [116]:
bank_account = 0
bank_account_list = []
for index in range(len(df_num_bank_accounts)):
    if (df_num_bank_accounts['Num_Bank_Accounts'][index]) > 50:
        bank_account +=1
        bank_account_list.append(df_num_bank_accounts['Num_Bank_Accounts'][index])
print("bank_account Count : ",bank_account,"|| Unique bank_account count :",len(set(bank_account_list))," || Example : ",bank_account_list[0:5])

bank_account Count :  1280 || Unique bank_account count : 909  || Example :  [1414, 1231, 67, 572, 1488]


In [119]:
for ssn in unique_SSN:
    corresponding_unique_bank_accounts_values[ssn] = list(df_num_bank_accounts[df_num_bank_accounts['SSN'] == ssn]['Num_Bank_Accounts'].unique())
dict(itertools.islice(corresponding_unique_bank_accounts_values.items(),90,100))

{87226137: [3],
 62293192: [2],
 727409990: [4],
 16732103: [6],
 198089791: [7],
 518012262: [9],
 465939571: [5],
 860361556: [3],
 407069232: [3],
 550948995: [6]}

replacing number of banks greater than 50

In [120]:
[val for val in corresponding_unique_bank_accounts_values[198089791] if val <50 ][0]

7

In [118]:
for index in range(len(df_num_bank_accounts)):
    if df_num_bank_accounts['Num_Bank_Accounts'][index] > 50:
        df_num_bank_accounts['Num_Bank_Accounts'][index] = [val for val in corresponding_unique_bank_accounts_values[df_num_bank_accounts['SSN'][index]] if val <= 50 ][0]

In [121]:
for ssn in unique_SSN:
    corresponding_unique_bank_accounts_values[ssn] = list(df_num_bank_accounts[df_num_bank_accounts['SSN'] == ssn]['Num_Bank_Accounts'].unique())
dict(itertools.islice(corresponding_unique_bank_accounts_values.items(),90,100))

{87226137: [3],
 62293192: [2],
 727409990: [4],
 16732103: [6],
 198089791: [7],
 518012262: [9],
 465939571: [5],
 860361556: [3],
 407069232: [3],
 550948995: [6]}

In [122]:
bank_account = 0
bank_account_list = []
for index in range(len(df_num_bank_accounts)):
    if (df_num_bank_accounts['Num_Bank_Accounts'][index]) > 50:
        bank_account +=1
        bank_account_list.append(df_num_bank_accounts['Num_Bank_Accounts'][index])
print("bank_account Count : ",bank_account,"|| Unique bank_account count :",len(set(bank_account_list))," || Example : ",bank_account_list[0:5])

bank_account Count :  0 || Unique bank_account count : 0  || Example :  []


### PreProcessing Num_Credit_Card

In [124]:
df_num_credit_card = df_num_bank_accounts.copy()

In [125]:
df_num_credit_card['Num_Credit_Card'].nunique()

1179

In [128]:
num_credit_card= 0
num_credit_card_list = []
for index in range(len(df_num_credit_card)):
    if (df_num_credit_card['Num_Credit_Card'][index]) > 50:
        num_credit_card += 1
        num_credit_card_list.append(df_num_credit_card['Num_Credit_Card'][index])
print("num_credit_cardCount : ",num_credit_card,"|| Unique num_credit_cardcount :",len(set(num_credit_card_list))," || Example : ",num_credit_card_list[0:5])

num_credit_cardCount :  2214 || Unique num_credit_cardcount : 1139  || Example :  [1385, 1288, 1029, 488, 1381]


In [129]:
corresponding_unique_num_credit_card_values = {}
for ssn in unique_SSN:
    corresponding_unique_num_credit_card_values[ssn] = list(df_num_credit_card[df_num_credit_card['SSN'] == ssn]['Num_Credit_Card'].unique())
dict(itertools.islice(corresponding_unique_num_credit_card_values.items(),10))

{821000265: [4],
 4075839: [4, 1385],
 486853974: [5],
 72316145: [5],
 615067821: [5],
 612708987: [1288, 5],
 411510676: [1],
 500926408: [7],
 70191622: [5],
 366681681: [5, 6]}

In [130]:
for index in range(len(df_num_credit_card)):
    if df_num_credit_card['Num_Credit_Card'][index] > 50:
        df_num_credit_card['Num_Credit_Card'][index] = [val for val in corresponding_unique_num_credit_card_values[df_num_credit_card['SSN'][index]] if val <= 50 ][0]

In [131]:
for ssn in unique_SSN:
    corresponding_unique_num_credit_card_values[ssn] = list(df_num_credit_card[df_num_credit_card['SSN'] == ssn]['Num_Credit_Card'].unique())
dict(itertools.islice(corresponding_unique_num_credit_card_values.items(),10))

{821000265: [4],
 4075839: [4],
 486853974: [5],
 72316145: [5],
 615067821: [5],
 612708987: [5],
 411510676: [1],
 500926408: [7],
 70191622: [5],
 366681681: [5, 6]}

In [132]:
num_credit_card= 0
num_credit_card_list = []
for index in range(len(df_num_credit_card)):
    if (df_num_credit_card['Num_Credit_Card'][index]) > 50:
        num_credit_card += 1
        num_credit_card_list.append(df_num_credit_card['Num_Credit_Card'][index])
print("num_credit_cardCount : ",num_credit_card,"|| Unique num_credit_cardcount :",len(set(num_credit_card_list))," || Example : ",num_credit_card_list[0:5])

num_credit_cardCount :  0 || Unique num_credit_cardcount : 0  || Example :  []


In [133]:
df_num_credit_card.to_csv('checkpoints//df_num_credit_card.csv',index = False)

### PreProcessing Interest_Rate

In [3]:
df_interest_rate = pd.read_csv('checkpoints//df_num_credit_card.csv')

In [4]:
df_interest_rate["Interest_Rate"].nunique()

1750

In [5]:
corresponding_unique_interest_rate_values = {}
for ssn in unique_SSN:
    corresponding_unique_interest_rate_values[ssn] = list(df_interest_rate[df_interest_rate['SSN'] == ssn]['Interest_Rate'].unique())

In [6]:
dict(itertools.islice(corresponding_unique_interest_rate_values.items(),10))

{821000265: [3],
 4075839: [6],
 486853974: [8],
 72316145: [4],
 615067821: [5],
 612708987: [8, 5318],
 411510676: [8],
 500926408: [15],
 70191622: [7],
 366681681: [12]}

In [7]:
interest_rate = 0
interest_rate_list = []
for index in range(len(df_interest_rate)):
    if (df_interest_rate['Interest_Rate'][index]) > 100:
        interest_rate +=1
        interest_rate_list.append(df_interest_rate['Interest_Rate'][index])
print("Interest_rate Count : ",interest_rate,"|| Unique interest_rate count :",len(set(interest_rate_list))," || Example : ",interest_rate_list[0:5])

Interest_rate Count :  2012 || Unique interest_rate count : 1697  || Example :  [5318, 433, 5240, 4975, 668]


In [8]:
for index in range(len(df_interest_rate)):
    if df_interest_rate['Interest_Rate'][index] > 100:
        df_interest_rate['Interest_Rate'][index] = [val for val in corresponding_unique_interest_rate_values[df_interest_rate['SSN'][index]] if val <= 100 ][0]

In [9]:
corresponding_unique_interest_rate_values = {}
for ssn in unique_SSN:
    corresponding_unique_interest_rate_values[ssn] = list(df_interest_rate[df_interest_rate['SSN'] == ssn]['Interest_Rate'].unique())

In [10]:
dict(itertools.islice(corresponding_unique_interest_rate_values.items(),10))

{821000265: [3],
 4075839: [6],
 486853974: [8],
 72316145: [4],
 615067821: [5],
 612708987: [8],
 411510676: [8],
 500926408: [15],
 70191622: [7],
 366681681: [12]}

In [12]:
interest_rate = 0
interest_rate_list = []
for index in range(len(df_interest_rate)):
    if (df_interest_rate['Interest_Rate'][index]) > 100:
        interest_rate +=1
        interest_rate_list.append(df_interest_rate['Interest_Rate'][index])
print("Interest_rate Count : ",interest_rate,"|| Unique interest_rate count :",len(set(interest_rate_list))," || Example : ",interest_rate_list[0:5])

Interest_rate Count :  0 || Unique interest_rate count : 0  || Example :  []


### PreProcessing Num_of_Loan

In [31]:
df_num_of_loan = df_interest_rate.copy()

In [32]:
df_num_of_loan['Num_of_Loan'].nunique()

434

In [33]:
underscore_num_of_loan = 0
underscore_num_of_loan_list = []
noisy_num_of_loan = 0
noisy_num_of_loan_list = []
for index in range(len(df_num_of_loan)):
    if '_' in  df_num_of_loan['Num_of_Loan'][index]:
        underscore_num_of_loan += 1
        underscore_num_of_loan_list.append(df_num_of_loan['Num_of_Loan'][index])
    try:
        if int(df_num_of_loan['Num_of_Loan'][index]) > 50 or int(df_num_of_loan['Num_of_Loan'][index] < 0):
            noisy_num_of_loan +=1
            noisy_num_of_loan_list.append(df_num_of_loan['Num_of_Loan'][index])
            
    except:
        pass
print("underscore_num_of_loan Count : ",underscore_num_of_loan,"|| Unique underscore_num_of_loan count :",len(set(underscore_num_of_loan_list)), " || Example : ",underscore_num_of_loan_list[0:5])
print("noisy_num_of_loan Count : ",noisy_num_of_loan,"|| Unique noisy_num_of_loan count :",len(set(noisy_num_of_loan_list))," || Example : ",noisy_num_of_loan_list[0:5])



underscore_num_of_loan Count :  4785 || Unique underscore_num_of_loan count : 35  || Example :  ['0_', '3_', '3_', '2_', '5_']
noisy_num_of_loan Count :  430 || Unique noisy_num_of_loan count : 376  || Example :  ['967', '1464', '622', '352', '472']


Ommiting underscores

In [34]:
for index in range(len(df_num_of_loan)):
    if '_' in df_num_of_loan['Num_of_Loan'][index]:
        df_num_of_loan['Num_of_Loan'][index] = df_num_of_loan['Num_of_Loan'][index].replace('_','')

Ommiting noisy datas

In [35]:
df_num_of_loan['Num_of_Loan'] = df_num_of_loan['Num_of_Loan'].astype(int)

In [36]:
corresponding_unique_num_of_loan_values= {}
for ssn in unique_SSN:
    corresponding_unique_num_of_loan_values[ssn] = list(df_num_of_loan[df_num_of_loan['SSN'] == ssn]['Num_of_Loan'].unique())

In [37]:
dict(itertools.islice(corresponding_unique_num_of_loan_values.items(),10))

{821000265: [4],
 4075839: [1],
 486853974: [3, 967],
 72316145: [1, -100],
 615067821: [0, -100],
 612708987: [0],
 411510676: [2, -100],
 500926408: [3, -100],
 70191622: [-100, 3],
 366681681: [2, -100]}

In [38]:
for index in range(len(df_num_of_loan)):
    if df_num_of_loan['Num_of_Loan'][index] > 50 or df_num_of_loan['Num_of_Loan'][index] < 0:
        df_num_of_loan['Num_of_Loan'][index] = [val for val in corresponding_unique_num_of_loan_values[df_num_of_loan['SSN'][index]] if ((val >= 0) and (val <= 50)) ][0]

In [39]:
corresponding_unique_num_of_loan_values= {}
for ssn in unique_SSN:
    corresponding_unique_num_of_loan_values[ssn] = list(df_num_of_loan[df_num_of_loan['SSN'] == ssn]['Num_of_Loan'].unique())

In [40]:
dict(itertools.islice(corresponding_unique_num_of_loan_values.items(),10))

{821000265: [4],
 4075839: [1],
 486853974: [3],
 72316145: [1],
 615067821: [0],
 612708987: [0],
 411510676: [2],
 500926408: [3],
 70191622: [3],
 366681681: [2]}

In [42]:
df_num_of_loan.to_csv('checkpoints//df_num_of_loan.csv',index = False)

### PreProcessing Type_of_Loan

In [3]:
df_type_of_loan = pd.read_csv('checkpoints//df_num_of_loan.csv')

In [44]:
df_type_of_loan['Type_of_Loan'].nunique()

6260

In [52]:
types_of_loan = set()

In [54]:
df_type_of_loan['Type_of_Loan'][10]

'Credit-Builder Loan'

In [74]:
df_type_of_loan['Type_of_Loan'][10].split(',')
df_type_of_loan['Type_of_Loan'][0].split(',')[2].split(' Loan')[0].replace(' ','').replace('and','')

['Credit-Builder Loan']

'Personal'

In [85]:
corresponding_unique_type_of_loan_values = {}
for ssn in unique_SSN:
    corresponding_unique_type_of_loan_values[ssn] = list(df_type_of_loan[df_type_of_loan['SSN'] == ssn]['Type_of_Loan'].unique())

In [86]:
dict(itertools.islice(corresponding_unique_type_of_loan_values.items(),10))

{821000265: ['Auto Loan, Credit-Builder Loan, Personal Loan, and Home Equity Loan'],
 4075839: ['Credit-Builder Loan'],
 486853974: ['Auto Loan, Auto Loan, and Not Specified'],
 72316145: ['Not Specified'],
 615067821: [nan],
 612708987: [nan],
 411510676: ['Credit-Builder Loan, and Mortgage Loan'],
 500926408: ['Not Specified, Auto Loan, and Student Loan'],
 70191622: ['Personal Loan, Debt Consolidation Loan, and Auto Loan'],
 366681681: ['Not Specified, and Payday Loan']}

replacing nan with 'Not Specified'

In [87]:
df_type_of_loan['Type_of_Loan'].fillna('Not Specified',inplace = True)

In [88]:
corresponding_unique_type_of_loan_values = {}
for ssn in unique_SSN:
    corresponding_unique_type_of_loan_values[ssn] = list(df_type_of_loan[df_type_of_loan['SSN'] == ssn]['Type_of_Loan'].unique())

In [101]:
dict(itertools.islice(corresponding_unique_type_of_loan_values.items(),20))

{821000265: ['Auto Loan, Credit-Builder Loan, Personal Loan, and Home Equity Loan'],
 4075839: ['Credit-Builder Loan'],
 486853974: ['Auto Loan, Auto Loan, and Not Specified'],
 72316145: ['Not Specified'],
 615067821: ['Not Specified'],
 612708987: ['Not Specified'],
 411510676: ['Credit-Builder Loan, and Mortgage Loan'],
 500926408: ['Not Specified, Auto Loan, and Student Loan'],
 70191622: ['Personal Loan, Debt Consolidation Loan, and Auto Loan'],
 366681681: ['Not Specified, and Payday Loan'],
 221308554: ['Credit-Builder Loan, Personal Loan, and Auto Loan'],
 342902649: ['Payday Loan, and Payday Loan'],
 414532918: ['Not Specified, Student Loan, and Personal Loan'],
 328336328: ['Personal Loan, Payday Loan, Student Loan, Auto Loan, Home Equity Loan, Student Loan, and Payday Loan'],
 655057666: ['Not Specified, Student Loan, Student Loan, Credit-Builder Loan, and Auto Loan'],
 965462491: ['Payday Loan, and Home Equity Loan'],
 891559364: ['Not Specified'],
 928914452: ['Credit-Buil

In [81]:
corresponding_unique_type_of_loan_values[821000265][0].split(',')

['Auto Loan',
 ' Credit-Builder Loan',
 ' Personal Loan',
 ' and Home Equity Loan']

In [90]:
for values in corresponding_unique_type_of_loan_values.values():
    for i in range(len(values)):
        types_of_loan.add(values[0].split(',')[i].split(' Loan')[0].replace(' ','').replace('and',''))

In [99]:
types_of_loan

{'Auto',
 'Credit-Builder',
 'DebtConsolidation',
 'HomeEquity',
 'Mortgage',
 'NotSpecified',
 'Payday',
 'Personal',
 'Student'}

In [141]:
processed_types_of_loan_list = sorted(list(types_of_loan))
processed_types_of_loan_list

['Auto',
 'Credit-Builder',
 'DebtConsolidation',
 'HomeEquity',
 'Mortgage',
 'NotSpecified',
 'Payday',
 'Personal',
 'Student']

In [143]:
dataframe_type_of_loan_list = sorted(['Credit-Builder Loan',
                               'Not Specified',
                               'Debt Consolidation Loan',
                               'Auto Loan',
                               'Student Loan',
                               'Home Equity Loan', 
                               'Mortgage Loan',
                               'Personal Loan',
                               'Payday Loan'])
dataframe_type_of_loan_list

['Auto Loan',
 'Credit-Builder Loan',
 'Debt Consolidation Loan',
 'Home Equity Loan',
 'Mortgage Loan',
 'Not Specified',
 'Payday Loan',
 'Personal Loan',
 'Student Loan']

In [144]:
def map_type_of_loan(txt):
    txt = str(txt)
    for i in range(len(processed_types_of_loan_list)):
        txt = txt.replace(dataframe_type_of_loan_list[i],processed_types_of_loan_list[i])
    return txt

In [145]:
txt = 'Credit-Builder Loan,Not Specified,Debt Consolidation Loan,Auto Loan,Student Loan,Home Equity Loan,Mortgage Loan,Personal Loan'
map_type_of_loan(txt)

'Credit-Builder,NotSpecified,DebtConsolidation,Auto,Student,HomeEquity,Mortgage,Personal'

In [139]:
df_type_of_loan['Type_of_Loan'] = df_type_of_loan['Type_of_Loan'].apply(map_type_of_loan)

In [140]:
df_type_of_loan['Type_of_Loan'].head()

0    Auto, Credit-Builder, Personal, and HomeEquity
1    Auto, Credit-Builder, Personal, and HomeEquity
2    Auto, Credit-Builder, Personal, and HomeEquity
3    Auto, Credit-Builder, Personal, and HomeEquity
4    Auto, Credit-Builder, Personal, and HomeEquity
Name: Type_of_Loan, dtype: object

In [173]:
index = 0
for type_of_loan in processed_types_of_loan_list:
    df_type_of_loan.insert(loc = 11 + index, column = type_of_loan , value = 0)
    index += 1

In [174]:
df_type_of_loan.drop(columns = 'NotSpecified',inplace = True)

In [175]:
df_type_of_loan.columns

Index(['Month', 'Age', 'SSN', 'Occupation', 'Annual_Income',
       'Monthly_Inhand_Salary', 'Num_Bank_Accounts', 'Num_Credit_Card',
       'Interest_Rate', 'Num_of_Loan', 'Type_of_Loan', 'Auto',
       'Credit-Builder', 'DebtConsolidation', 'HomeEquity', 'Mortgage',
       'Payday', 'Personal', 'Student', 'Delay_from_due_date',
       'Num_of_Delayed_Payment', 'Changed_Credit_Limit',
       'Num_Credit_Inquiries', 'Credit_Mix', 'Outstanding_Debt',
       'Credit_Utilization_Ratio', 'Credit_History_Age',
       'Payment_of_Min_Amount', 'Total_EMI_per_month',
       'Amount_invested_monthly', 'Payment_Behaviour', 'Monthly_Balance',
       'Credit_Score'],
      dtype='object')

In [176]:
for type_of_loan in processed_types_of_loan_list:    
    df_type_of_loan[type_of_loan] = df_type_of_loan['Type_of_Loan'].apply(lambda x:1 if type_of_loan in x else 0)

In [178]:
df_type_of_loan.to_csv('checkpoints//df_type_of_loan.csv',index = False)

### PreProcessing Delay_from_due_date

In [4]:
df_delay_from_due_date = pd.read_csv('checkpoints//df_type_of_loan.csv')

In [5]:
df_delay_from_due_date['Delay_from_due_date'].isna().sum()

0

In [6]:
df_delay_from_due_date['Delay_from_due_date'].head()

0    3
1   -1
2    3
3    5
4    6
Name: Delay_from_due_date, dtype: int64

In [47]:
df_delay_from_due_date["Delay_from_due_date"].nunique()

73

In [7]:
corresponding_delay_from_due_date_values = {}

In [8]:
for ssn in unique_SSN:
    corresponding_delay_from_due_date_values[ssn] = list(df_delay_from_due_date[df_delay_from_due_date['SSN'] == ssn]['Delay_from_due_date'].unique())

In [9]:
dict(itertools.islice(corresponding_delay_from_due_date_values.items(),10))

{821000265: [3, -1, 5, 6, 8],
 4075839: [3, 7],
 486853974: [5, 13, 8, 10],
 72316145: [0, 5, 3, 7, 4],
 615067821: [5, 9, 1, 10],
 612708987: [12, 8, 11, 7],
 411510676: [0, -1],
 500926408: [30, 31, 34, 27],
 70191622: [13, 14, 11],
 366681681: [4, 2, -2, 1, -1]}

Seems random so not replacing probably dropping later.

### PreProcessing Num_of_Delayed_Payment

In [10]:
df_num_of_delayed_payment = df_delay_from_due_date.copy()

In [11]:
df_num_of_delayed_payment["Num_of_Delayed_Payment"].nunique()

749

In [12]:
df_num_of_delayed_payment["Num_of_Delayed_Payment"].isna().sum()

7002

In [14]:
corresponding_unique_delayed_payment_values = {}

In [15]:
for ssn in unique_SSN:
    corresponding_unique_delayed_payment_values[ssn] = list(df_num_of_delayed_payment[df_num_of_delayed_payment['SSN'] == ssn]['Num_of_Delayed_Payment'].unique())

In [18]:
dict(itertools.islice(corresponding_unique_delayed_payment_values.items(),10))

{821000265: ['7', nan, '4', '8_', '6'],
 4075839: ['4', '1', '-1', '3_', '0'],
 486853974: ['8', '6', '7', '5'],
 72316145: ['6', '3', '9', nan],
 615067821: [nan, '12', '15', '17'],
 612708987: ['10', '7', '4'],
 411510676: ['3', '2', '4', '3_', '2_'],
 500926408: ['11', '14'],
 70191622: ['11', '8', '14', nan],
 366681681: ['0', '2']}

Making values int

In [21]:
underscore_unique_delayed_payment = 0
underscore_unique_delayed_payment_list = []
noisy_unique_delayed_payment = 0
noisy_unique_delayed_payment_list = []
for index in range(len(df_num_of_delayed_payment)):
    try:
        if '_' in  df_num_of_delayed_payment['Num_of_Delayed_Payment'][index]:
            underscore_unique_delayed_payment += 1
            underscore_unique_delayed_payment_list.append(df_num_of_delayed_payment['Num_of_Delayed_Payment'][index])
    except:
        pass
    try:
        if int(df_num_of_delayed_payment['Num_of_Delayed_Payment'][index]) > 120 or int(df_num_of_delayed_payment['Num_of_Delayed_Payment'][index] < 0):
            noisy_unique_delayed_payment +=1
            noisy_unique_delayed_payment_list.append(df_num_of_delayed_payment['Num_of_Delayed_Payment'][index])
            
    except:
        pass
print("underscore_unique_delayed_payment Count : ",underscore_unique_delayed_payment,"|| Unique underscore_unique_delayed_payment count :",len(set(underscore_unique_delayed_payment_list)), " || Example : ",underscore_unique_delayed_payment_list[0:5])
print("Noisy_unique_delayed_payment Count : ",noisy_unique_delayed_payment,"|| Unique Noisy_unique_delayed_payment count :",len(set(noisy_unique_delayed_payment_list))," || Example : ",noisy_unique_delayed_payment_list[0:5])



underscore_unique_delayed_payment Count :  2744 || Unique underscore_unique_delayed_payment count : 51  || Example :  ['8_', '3_', '3_', '2_', '13_']
Noisy_unique_delayed_payment Count :  702 || Unique Noisy_unique_delayed_payment count : 651  || Example :  ['3318', '3083', '1338', '3104', '1106']


Noisy data maybe dropped later. For now replacing underscore value

In [22]:
for index in range(len(df_num_of_delayed_payment)):
    try:
        if '_' in df_num_of_delayed_payment['Num_of_Delayed_Payment'][index]:
            df_num_of_delayed_payment['Num_of_Delayed_Payment'][index] = df_num_of_delayed_payment['Num_of_Delayed_Payment'][index].replace('_','')
    except:
        pass

In [25]:
df_num_of_delayed_payment['Num_of_Delayed_Payment'] = df_num_of_delayed_payment['Num_of_Delayed_Payment'].astype(float)


In [26]:
underscore_unique_delayed_payment = 0
underscore_unique_delayed_payment_list = []
noisy_unique_delayed_payment = 0
noisy_unique_delayed_payment_list = []
for index in range(len(df_num_of_delayed_payment)):
    try:
        if '_' in  df_num_of_delayed_payment['Num_of_Delayed_Payment'][index]:
            underscore_unique_delayed_payment += 1
            underscore_unique_delayed_payment_list.append(df_num_of_delayed_payment['Num_of_Delayed_Payment'][index])
    except:
        pass
    try:
        if int(df_num_of_delayed_payment['Num_of_Delayed_Payment'][index]) > 120 or int(df_num_of_delayed_payment['Num_of_Delayed_Payment'][index] < 0):
            noisy_unique_delayed_payment +=1
            noisy_unique_delayed_payment_list.append(df_num_of_delayed_payment['Num_of_Delayed_Payment'][index])
            
    except:
        pass
print("underscore_unique_delayed_payment Count : ",underscore_unique_delayed_payment,"|| Unique underscore_unique_delayed_payment count :",len(set(underscore_unique_delayed_payment_list)), " || Example : ",underscore_unique_delayed_payment_list[0:5])
print("Noisy_unique_delayed_payment Count : ",noisy_unique_delayed_payment,"|| Unique Noisy_unique_delayed_payment count :",len(set(noisy_unique_delayed_payment_list))," || Example : ",noisy_unique_delayed_payment_list[0:5])



underscore_unique_delayed_payment Count :  0 || Unique underscore_unique_delayed_payment count : 0  || Example :  []
Noisy_unique_delayed_payment Count :  1365 || Unique Noisy_unique_delayed_payment count : 667  || Example :  [-1.0, 3318.0, 3083.0, 1338.0, 3104.0]


it seems after removing underscore some datas became noisy

### PreProcessing Changed_Credit_Limit

In [49]:
df_changed_credit_limit = df_num_of_delayed_payment.copy()

In [50]:
df_changed_credit_limit['Changed_Credit_Limit'].nunique()

4384

In [51]:
df_changed_credit_limit['Changed_Credit_Limit'].isna().sum()

0

In [52]:
corresponding_unique_changed_credit_limit_values = {}

In [53]:
for ssn in unique_SSN:
    corresponding_unique_changed_credit_limit_values[ssn] = list(df_changed_credit_limit[df_changed_credit_limit['SSN'] == ssn]['Changed_Credit_Limit'].unique())

In [54]:
dict(itertools.islice(corresponding_unique_changed_credit_limit_values.items(),10))

{821000265: ['11.27', '_', '6.27', '9.27'],
 4075839: ['5.42', '7.42', '6.42'],
 486853974: ['7.1', '11.1', '9.1'],
 72316145: ['1.99', '-2.01', '-1.01', '-3.01'],
 615067821: ['2.58'],
 612708987: ['10.14', '9.14'],
 411510676: ['9.34', '15.34', '8.34', '11.34'],
 500926408: ['17.13', '21.13', '18.13'],
 70191622: ['8.24', '_', '11.24'],
 366681681: ['5.76', '-1.2400000000000002', '1.7599999999999998']}

In [55]:
underscore_changed_credit_limit = 0
underscore_changed_credit_limit_list = []
noisy_changed_credit_limit = 0
noisy_changed_credit_limit_list = []
for index in range(len(df_changed_credit_limit)):
    if '_' in  df_changed_credit_limit['Changed_Credit_Limit'][index]:
        underscore_changed_credit_limit += 1
        underscore_changed_credit_limit_list.append(df_changed_credit_limit['Changed_Credit_Limit'][index])
    try:
        if int(df_changed_credit_limit['Changed_Credit_Limit'][index]) > 120 or int(df_changed_credit_limit['Changed_Credit_Limit'][index] <= 0):
            noisy_changed_credit_limit +=1
            noisy_changed_credit_limit_list.append(df_changed_credit_limit['Changed_Credit_Limit'][index])
            
    except:
        pass
print("underscore_changed_credit_limit Count : ",underscore_changed_credit_limit,"|| Unique underscore_changed_credit_limit count :",len(set(underscore_changed_credit_limit_list)), " || Example : ",underscore_changed_credit_limit_list[0:5])
print("noisy_changed_credit_limit Count : ",noisy_changed_credit_limit,"|| Unique noisy_changed_credit_limit count :",len(set(noisy_changed_credit_limit_list))," || Example : ",noisy_changed_credit_limit_list[0:5])



underscore_changed_credit_limit Count :  2091 || Unique underscore_changed_credit_limit count : 1  || Example :  ['_', '_', '_', '_', '_']
noisy_changed_credit_limit Count :  0 || Unique noisy_changed_credit_limit count : 0  || Example :  []


In [56]:
for index in range(len(df_changed_credit_limit)):
    if '_' in df_changed_credit_limit['Changed_Credit_Limit'][index]:
        df_changed_credit_limit['Changed_Credit_Limit'][index] = df_changed_credit_limit['Changed_Credit_Limit'][index].replace('_','nan')

In [57]:
df_changed_credit_limit['Changed_Credit_Limit'] = df_changed_credit_limit['Changed_Credit_Limit'].astype(float)


In [59]:
underscore_changed_credit_limit = 0
underscore_changed_credit_limit_list = []
noisy_changed_credit_limit = 0
noisy_changed_credit_limit_list = []
for index in range(len(df_changed_credit_limit)):
    if float(df_changed_credit_limit['Changed_Credit_Limit'][index]) > 120 or int(df_changed_credit_limit['Changed_Credit_Limit'][index] <= 0):
        noisy_changed_credit_limit +=1
        noisy_changed_credit_limit_list.append(df_changed_credit_limit['Changed_Credit_Limit'][index])

print("underscore_changed_credit_limit Count : ",underscore_changed_credit_limit,"|| Unique underscore_changed_credit_limit count :",len(set(underscore_changed_credit_limit_list)), " || Example : ",underscore_changed_credit_limit_list[0:5])
print("noisy_changed_credit_limit Count : ",noisy_changed_credit_limit,"|| Unique noisy_changed_credit_limit count :",len(set(noisy_changed_credit_limit_list))," || Example : ",noisy_changed_credit_limit_list[0:5])



underscore_changed_credit_limit Count :  0 || Unique underscore_changed_credit_limit count : 0  || Example :  []
noisy_changed_credit_limit Count :  1590 || Unique noisy_changed_credit_limit count : 661  || Example :  [-2.01, -1.01, -3.01, -1.2400000000000002, -4.14]


Also seems random so leaving as it is for now

### PreProcessing Num_Credit_Inquiries

In [60]:
df_num_credit_inquires = df_changed_credit_limit.copy()

In [61]:
df_num_credit_inquires['Num_Credit_Inquiries'].nunique()

1223

In [62]:
df_num_credit_inquires['Num_Credit_Inquiries'].isna().sum()

1965

In [63]:
corresponding_unique_num_credit_inquires_values = {}

In [64]:
for ssn in unique_SSN:
    corresponding_unique_num_credit_inquires_values[ssn] = list(df_num_credit_inquires[df_num_credit_inquires['SSN'] == ssn]['Num_Credit_Inquiries'].unique())

In [65]:
dict(itertools.islice(corresponding_unique_num_credit_inquires_values.items(),10))

{821000265: [4.0],
 4075839: [2.0],
 486853974: [3.0, nan],
 72316145: [4.0],
 615067821: [4.0],
 612708987: [2.0, nan],
 411510676: [2.0, 4.0],
 500926408: [5.0, 9.0],
 70191622: [3.0, 8.0],
 366681681: [2.0, 5.0]}

In [67]:
underscore_num_credit_inquires = 0
underscore_num_credit_inquires_list = []
noisy_num_credit_inquires = 0
noisy_num_credit_inquires_list = []
for index in range(len(df_num_credit_inquires)):
    try:
        if '_' in  df_num_credit_inquires['Num_Credit_Inquiries'][index]:
            underscore_num_credit_inquires += 1
            underscore_num_credit_inquires_list.append(df_num_credit_inquires['Num_Credit_Inquiries'][index])
    except:
        pass
    try:
        if int(df_num_credit_inquires['Num_Credit_Inquiries'][index]) > 120 or int(df_num_credit_inquires['Num_Credit_Inquiries'][index] <= 0):
            noisy_num_credit_inquires +=1
            noisy_num_credit_inquires_list.append(df_num_credit_inquires['Num_Credit_Inquiries'][index])
            
    except:
        pass
print("underscore_num_credit_inquires Count : ",underscore_num_credit_inquires,"|| Unique underscore_num_credit_inquires count :",len(set(underscore_num_credit_inquires_list)), " || Example : ",underscore_num_credit_inquires_list[0:5])
print("noisy_num_credit_inquires Count : ",noisy_num_credit_inquires,"|| Unique noisy_num_credit_inquires count :",len(set(noisy_num_credit_inquires_list))," || Example : ",noisy_num_credit_inquires_list[0:5])



underscore_num_credit_inquires Count :  0 || Unique underscore_num_credit_inquires count : 0  || Example :  []
noisy_num_credit_inquires Count :  8562 || Unique noisy_num_credit_inquires count : 1164  || Example :  [0.0, 0.0, 0.0, 1050.0, 0.0]


In [70]:
df_num_credit_inquires.to_csv('checkpoints//df_credit_inquires.csv',index = False)

leaving it for now

### PreProcessing Credit_Mix

In [80]:
df_credit_mix = pd.read_csv('checkpoints//df_credit_inquires.csv')

In [81]:
df_credit_mix['Credit_Mix'].nunique()

4

In [82]:
df_credit_mix['Credit_Mix'].value_counts()

Standard    36479
Good        24337
_           20195
Bad         18989
Name: Credit_Mix, dtype: int64

In [83]:
corresponding_unique_credit_mix_values = {}

In [84]:
for ssn in unique_SSN:
    corresponding_unique_credit_mix_values[ssn] = list(df_credit_mix[df_credit_mix['SSN'] == ssn]['Credit_Mix'].unique())

In [85]:
dict(itertools.islice(corresponding_unique_credit_mix_values.items(),10))

{821000265: ['_', 'Good'],
 4075839: ['Good', '_'],
 486853974: ['Good', '_'],
 72316145: ['Good', '_'],
 615067821: ['Standard', '_'],
 612708987: ['Good', '_'],
 411510676: ['Good'],
 500926408: ['Standard', '_'],
 70191622: ['Good', '_'],
 366681681: ['Good', '_']}

In [88]:
for index in range(len(df_credit_mix)):
    if '_' in df_credit_mix['Credit_Mix'][index]:
        df_credit_mix['Credit_Mix'][index] = [val for val in corresponding_unique_credit_mix_values[df_credit_mix['SSN'][index]] if not '_' in val][0]

In [89]:
corresponding_unique_credit_mix_values = {}

In [90]:
for ssn in unique_SSN:
    corresponding_unique_credit_mix_values[ssn] = list(df_credit_mix[df_credit_mix['SSN'] == ssn]['Credit_Mix'].unique())

In [91]:
dict(itertools.islice(corresponding_unique_credit_mix_values.items(),10))

{821000265: ['Good'],
 4075839: ['Good'],
 486853974: ['Good'],
 72316145: ['Good'],
 615067821: ['Standard'],
 612708987: ['Good'],
 411510676: ['Good'],
 500926408: ['Standard'],
 70191622: ['Good'],
 366681681: ['Good']}

In [92]:
df_credit_mix['Credit_Mix'].value_counts()

Standard    45848
Good        30384
Bad         23768
Name: Credit_Mix, dtype: int64

### PreProcessing Outstanding_Debt

In [93]:
df_outstanding_debt = df_credit_mix.copy()

In [100]:
df_outstanding_debt['Outstanding_Debt'].nunique()

13178

In [94]:
corresponding_unique_outstanding_debt_values = {}

In [95]:
for ssn in unique_SSN:
    corresponding_unique_outstanding_debt_values[ssn] = list(df_outstanding_debt[df_outstanding_debt['SSN'] == ssn]['Outstanding_Debt'].unique())

In [99]:
dict(itertools.islice(corresponding_unique_outstanding_debt_values.items(),15))

{821000265: ['809.98'],
 4075839: ['605.03'],
 486853974: ['1303.01'],
 72316145: ['632.46'],
 615067821: ['943.86'],
 612708987: ['548.2'],
 411510676: ['352.16'],
 500926408: ['1704.18'],
 70191622: ['1377.74'],
 366681681: ['421.43'],
 221308554: ['1328.93', '1328.93_'],
 342902649: ['950.36'],
 414532918: ['179.22'],
 328336328: ['2602.69'],
 655057666: ['758.44']}

Replacing the underscore value

In [102]:
underscore_outstanding_debt = 0
underscore_outstanding_debt_list = []
for index in range(len(df_outstanding_debt)):
    if '_' in  df_outstanding_debt['Outstanding_Debt'][index]:
        underscore_outstanding_debt += 1
        underscore_outstanding_debt_list.append(df_outstanding_debt['Outstanding_Debt'][index])
print("underscore_outstanding_debt Count : ",underscore_outstanding_debt,"|| Unique underscore_outstanding_debt count :",len(set(underscore_outstanding_debt_list)), " || Example : ",underscore_outstanding_debt_list[0:5])

underscore_outstanding_debt Count :  1009 || Unique underscore_outstanding_debt count : 975  || Example :  ['1328.93_', '1283.37_', '2797.17_', '3818.57_', '343.84_']


In [104]:
for index in range(len(df_outstanding_debt)):
    if '_' in df_outstanding_debt['Outstanding_Debt'][index]:
        df_outstanding_debt['Outstanding_Debt'][index] = [val for val in corresponding_unique_outstanding_debt_values[df_outstanding_debt['SSN'][index]] if not '_' in val][0]

In [105]:
df_outstanding_debt['Outstanding_Debt'] = df_outstanding_debt['Outstanding_Debt'].astype(float)

In [107]:
corresponding_unique_outstanding_debt_values = {}
for ssn in unique_SSN:
    corresponding_unique_outstanding_debt_values[ssn] = list(df_outstanding_debt[df_outstanding_debt['SSN'] == ssn]['Outstanding_Debt'].unique())

In [110]:
dict(itertools.islice(corresponding_unique_outstanding_debt_values.items(),10))

{821000265: [809.98],
 4075839: [605.03],
 486853974: [1303.01],
 72316145: [632.46],
 615067821: [943.86],
 612708987: [548.2],
 411510676: [352.16],
 500926408: [1704.18],
 70191622: [1377.74],
 366681681: [421.43]}

all customer seems to have same outstanding debt

### PreProcessing Credit_Utilization_Ratio

In [175]:
df_credit_utilization_ratio = df_outstanding_debt.copy()

In [176]:
df_credit_utilization_ratio["Credit_Utilization_Ratio"].isna().sum()

0

In [177]:
corresponding_unique_credit_utilization_ratio_values = {}
for ssn in unique_SSN:
    corresponding_unique_credit_utilization_ratio_values[ssn] = list(df_credit_utilization_ratio[df_credit_utilization_ratio['SSN'] == ssn]['Credit_Utilization_Ratio'].unique())

In [178]:
dict(itertools.islice(corresponding_unique_credit_utilization_ratio_values.items(),3))

{821000265: [26.822619623699016,
  31.94496005538421,
  28.60935202206993,
  31.37786186958236,
  24.797346908844982,
  27.26225871052017,
  22.53759303178384,
  23.93379480196552],
 4075839: [24.46403063758457,
  38.550848433956325,
  33.22495078663659,
  39.18265565546884,
  34.97789474709241,
  33.38101020065065,
  31.131701611553307,
  32.933856293622235],
 486853974: [28.616734823006844,
  41.70257342393906,
  26.51981539154649,
  39.50164810812735,
  31.37614957421204,
  39.78399345802119,
  38.068623627698585,
  38.37475294778511]}

Each customer has different credit_utilization_ratio

### PreProcessing Credit_History_Age

In [233]:
df_credit_history_age = df_credit_utilization_ratio.copy()

In [234]:
df_credit_history_age['Credit_History_Age'].nunique()

404

In [235]:
df_credit_history_age['Credit_History_Age'].isna().sum()

9030

In [236]:
df_credit_history_age[0:10]

Month  Age        SSN Occupation  Annual_Income  Monthly_Inhand_Salary  \
0   January   23  821000265  Scientist       19114.12            1824.843333   
1  February   23  821000265  Scientist       19114.12            1824.843333   
2     March   23  821000265  Scientist       19114.12            1824.843333   
3     April   23  821000265  Scientist       19114.12            1824.843333   
4       May   23  821000265  Scientist       19114.12            1824.843333   
5      June   23  821000265  Scientist       19114.12            1824.843333   
6      July   23  821000265  Scientist       19114.12            1824.843333   
7    August   23  821000265  Scientist       19114.12            1824.843333   
8   January   28    4075839    Teacher       34847.84            3037.986667   
9  February   28    4075839    Teacher       34847.84            3037.986667   

   Num_Bank_Accounts  Num_Credit_Card  Interest_Rate  Num_of_Loan  ...  \
0                  3                4              3            4  ...   
1                  3                4              3            4  ...   
2                  3                4              3            4  ...   
3                  3                4              3            4  ...   
4                  3                4              3            4  ...   
5                  3                4              3            4  ...   
6                  3                4              3            4  ...   
7                  3                4              3            4  ...   
8                  2                4              6            1  ...   
9                  2                4              6            1  ...   

  Outstanding_Debt  Credit_Utilization_Ratio     Credit_History_Age  \
0           809.98                 26.822620  22 Years and 1 Months   
1           809.98                 31.944960                    NaN   
2           809.98                 28.609352  22 Years and 3 Months   
3           809.98                 31.377862  22 Years and 4 Months   
4           809.98                 24.797347  22 Years and 5 Months   
5           809.98                 27.262259  22 Years and 6 Months   
6           809.98                 22.537593  22 Years and 7 Months   
7           809.98                 23.933795                    NaN   
8           605.03                 24.464031  26 Years and 7 Months   
9           605.03                 38.550848  26 Years and 8 Months   

   Payment_of_Min_Amount  Total_EMI_per_month  Amount_invested_monthly  \
0                     No            49.574949        80.41529543900253   
1                     No            49.574949       118.28022162236736   
2                     No            49.574949          81.699521264648   
3                     No            49.574949        199.4580743910713   
4                     No            49.574949       41.420153086217326   
5                     No            49.574949       62.430172331195294   
6                     No            49.574949        178.3440674122349   
7                     No            49.574949       24.785216509052056   
8                     No            18.816215         104.291825168246   
9                     No            18.816215        40.39123782853101   

                  Payment_Behaviour     Monthly_Balance  Credit_Score  \
0   High_spent_Small_value_payments  312.49408867943663          Good   
1    Low_spent_Large_value_payments  284.62916249607184          Good   
2   Low_spent_Medium_value_payments   331.2098628537912          Good   
3    Low_spent_Small_value_payments  223.45130972736786          Good   
4  High_spent_Medium_value_payments  341.48923103222177          Good   
5                            !@9#%8   340.4792117872438          Good   
6    Low_spent_Small_value_payments   244.5653167062043          Good   
7  High_spent_Medium_value_payments  358.12416760938714      Standard   
8    Low_spent_Small_value_payments  470.69062692529184 

Each row seems to be increased by 1 month

In [237]:
def format_year(txt):
    return txt.replace(' ','_')

In [238]:
df_credit_history_age['Credit_History_Age'].fillna('nan',inplace = True)

In [239]:
df_credit_history_age['Credit_History_Age'] = df_credit_history_age['Credit_History_Age'].apply(format_year) 

In [240]:
df_credit_history_age.head()

Month  Age        SSN Occupation  Annual_Income  Monthly_Inhand_Salary  \
0   January   23  821000265  Scientist       19114.12            1824.843333   
1  February   23  821000265  Scientist       19114.12            1824.843333   
2     March   23  821000265  Scientist       19114.12            1824.843333   
3     April   23  821000265  Scientist       19114.12            1824.843333   
4       May   23  821000265  Scientist       19114.12            1824.843333   

   Num_Bank_Accounts  Num_Credit_Card  Interest_Rate  Num_of_Loan  ...  \
0                  3                4              3            4  ...   
1                  3                4              3            4  ...   
2                  3                4              3            4  ...   
3                  3                4              3            4  ...   
4                  3                4              3            4  ...   

  Outstanding_Debt  Credit_Utilization_Ratio     Credit_History_Age  \
0           809.98                 26.822620  22_Years_and_1_Months   
1           809.98                 31.944960                    nan   
2           809.98                 28.609352  22_Years_and_3_Months   
3           809.98                 31.377862  22_Years_and_4_Months   
4           809.98                 24.797347  22_Years_and_5_Months   

   Payment_of_Min_Amount  Total_EMI_per_month  Amount_invested_monthly  \
0                     No            49.574949        80.41529543900253   
1                     No            49.574949       118.28022162236736   
2                     No            49.574949          81.699521264648   
3                     No            49.574949        199.4580743910713   
4                     No            49.574949       41.420153086217326   

                  Payment_Behaviour     Monthly_Balance  Credit_Score  \
0   High_spent_Small_value_payments  312.49408867943663          Good   
1    Low_spent_Large_value_payments  284.62916249607184          Good   
2   Low_spent_Medium_value_payments   331.2098628537912          Good   
3    Low_spent_Small_value_payments  223.45130972736786          Good   
4  High_spent_Medium_value_payments  341.48923103222177          Good   

   NotSpecified  
0             0  
1             0  
2             0  
3             0  
4             0  

[5 rows x 34 columns]

In [241]:
check = '22_Years_and_1_Months'
int(check.split('_Years')[0])
int(check.split('and_')[1].split('_Months')[0])

22

1

In [242]:
for index in range(1,len(df_credit_history_age)):
    if df_credit_history_age['Credit_History_Age'][index] == 'nan':
        if df_credit_history_age['SSN'][index] == df_credit_history_age['SSN'][index - 1] and df_credit_history_age['Credit_History_Age'][index - 1] != 'nan':
            previous_year = int(df_credit_history_age['Credit_History_Age'][index - 1].split('_Years')[0])
            previous_month = int(df_credit_history_age['Credit_History_Age'][index - 1].split('and_')[1].split('_Months')[0])
            
            if previous_month == 11:
                df_credit_history_age['Credit_History_Age'][index] = f'{previous_year + 1}_Years_and_0_Months'
            else:
                df_credit_history_age['Credit_History_Age'][index] = f'{previous_year}_Years_and_{previous_month + 1}_Months'
        
        
        elif df_credit_history_age['SSN'][index] == df_credit_history_age['SSN'][index + 1] and df_credit_history_age['Credit_History_Age'][index + 1] != 'nan':
            upcoming_year = int(df_credit_history_age['Credit_History_Age'][index + 1].split('_Years')[0])
            upcoming_month = int(df_credit_history_age['Credit_History_Age'][index + 1].split('and_')[1].split('_Months')[0])
            
            if upcoming_month == 0:
                df_credit_history_age['Credit_History_Age'][index] = f'{upcoming_year - 1}_Years_and_11_Months'
            else:
                df_credit_history_age['Credit_History_Age'][index] = f'{upcoming_year}_Years_and_{upcoming_month - 1}_Months'
        
        elif df_credit_history_age['SSN'][index] == df_credit_history_age['SSN'][index + 2] and df_credit_history_age['Credit_History_Age'][index + 2] != 'nan':
            upcoming_year = int(df_credit_history_age['Credit_History_Age'][index + 2].split('_Years')[0])
            upcoming_month = int(df_credit_history_age['Credit_History_Age'][index + 2].split('and_')[1].split('_Months')[0])
            
            if upcoming_month == 1:
                df_credit_history_age['Credit_History_Age'][index] = f'{upcoming_year - 1}_Years_and_11_Months'
            else:
                df_credit_history_age['Credit_History_Age'][index] = f'{upcoming_year}_Years_and_{upcoming_month - 2}_Months'
        
        elif df_credit_history_age['SSN'][index] == df_credit_history_age['SSN'][index - 1] and df_credit_history_age['Credit_History_Age'][index - 1] != 'nan':
            previous_year = int(df_credit_history_age['Credit_History_Age'][index - 2].split('_Years')[0])
            previous_month = int(df_credit_history_age['Credit_History_Age'][index - 2].split('and_')[1].split('_Months')[0])
            
            if previous_month == 10:
                df_credit_history_age['Credit_History_Age'][index] = f'{previous_year + 1}_Years_and_0_Months'
            else:
                df_credit_history_age['Credit_History_Age'][index] = f'{previous_year}_Years_and_{previous_month + 2}_Months'
        

In [243]:
df_credit_history_age[0:10]

Month  Age        SSN Occupation  Annual_Income  Monthly_Inhand_Salary  \
0   January   23  821000265  Scientist       19114.12            1824.843333   
1  February   23  821000265  Scientist       19114.12            1824.843333   
2     March   23  821000265  Scientist       19114.12            1824.843333   
3     April   23  821000265  Scientist       19114.12            1824.843333   
4       May   23  821000265  Scientist       19114.12            1824.843333   
5      June   23  821000265  Scientist       19114.12            1824.843333   
6      July   23  821000265  Scientist       19114.12            1824.843333   
7    August   23  821000265  Scientist       19114.12            1824.843333   
8   January   28    4075839    Teacher       34847.84            3037.986667   
9  February   28    4075839    Teacher       34847.84            3037.986667   

   Num_Bank_Accounts  Num_Credit_Card  Interest_Rate  Num_of_Loan  ...  \
0                  3                4              3            4  ...   
1                  3                4              3            4  ...   
2                  3                4              3            4  ...   
3                  3                4              3            4  ...   
4                  3                4              3            4  ...   
5                  3                4              3            4  ...   
6                  3                4              3            4  ...   
7                  3                4              3            4  ...   
8                  2                4              6            1  ...   
9                  2                4              6            1  ...   

  Outstanding_Debt  Credit_Utilization_Ratio     Credit_History_Age  \
0           809.98                 26.822620  22_Years_and_1_Months   
1           809.98                 31.944960  22_Years_and_2_Months   
2           809.98                 28.609352  22_Years_and_3_Months   
3           809.98                 31.377862  22_Years_and_4_Months   
4           809.98                 24.797347  22_Years_and_5_Months   
5           809.98                 27.262259  22_Years_and_6_Months   
6           809.98                 22.537593  22_Years_and_7_Months   
7           809.98                 23.933795  22_Years_and_8_Months   
8           605.03                 24.464031  26_Years_and_7_Months   
9           605.03                 38.550848  26_Years_and_8_Months   

   Payment_of_Min_Amount  Total_EMI_per_month  Amount_invested_monthly  \
0                     No            49.574949        80.41529543900253   
1                     No            49.574949       118.28022162236736   
2                     No            49.574949          81.699521264648   
3                     No            49.574949        199.4580743910713   
4                     No            49.574949       41.420153086217326   
5                     No            49.574949       62.430172331195294   
6                     No            49.574949        178.3440674122349   
7                     No            49.574949       24.785216509052056   
8                     No            18.816215         104.291825168246   
9                     No            18.816215        40.39123782853101   

                  Payment_Behaviour     Monthly_Balance  Credit_Score  \
0   High_spent_Small_value_payments  312.49408867943663          Good   
1    Low_spent_Large_value_payments  284.62916249607184          Good   
2   Low_spent_Medium_value_payments   331.2098628537912          Good   
3    Low_spent_Small_value_payments  223.45130972736786          Good   
4  High_spent_Medium_value_payments  341.48923103222177          Good   
5                            !@9#%8   340.4792117872438          Good   
6    Low_spent_Small_value_payments   244.5653167062043          Good   
7  High_spent_Medium_value_payments  358.12416760938714      Standard   
8    Low_spent_Small_value_payments  470.69062692529184 

In [244]:
df_credit_history_age.to_csv('checkpoints//df_credit_history_age.csv',index = False)

### PreProcessing Payment_of_Min_Amount

In [245]:
df_payment_of_min_amount = pd.read_csv('checkpoints//df_credit_history_age.csv')

In [246]:
df_payment_of_min_amount['Payment_of_Min_Amount'].nunique()

3

In [247]:
df_payment_of_min_amount['Payment_of_Min_Amount'].value_counts()

Yes    52326
No     35667
NM     12007
Name: Payment_of_Min_Amount, dtype: int64

In [248]:
corresponding_payment_of_min_values = {}
for ssn in unique_SSN:
    corresponding_payment_of_min_values[ssn] = list(df_payment_of_min_amount[df_payment_of_min_amount['SSN'] == ssn]['Payment_of_Min_Amount'].unique())

In [249]:
dict(itertools.islice(corresponding_payment_of_min_values.items(),10))

{821000265: ['No'],
 4075839: ['No', 'NM'],
 486853974: ['No'],
 72316145: ['No', 'NM'],
 615067821: ['Yes', 'NM'],
 612708987: ['No', 'NM'],
 411510676: ['NM', 'No'],
 500926408: ['NM', 'Yes'],
 70191622: ['No', 'NM'],
 366681681: ['NM', 'No']}

Replacing NM with corresponding value

In [250]:
for index in range(len(df_payment_of_min_amount)):
    if df_payment_of_min_amount['Payment_of_Min_Amount'][index] == 'NM':
        df_payment_of_min_amount['Payment_of_Min_Amount'][index] = [val for val in corresponding_payment_of_min_values[df_payment_of_min_amount['SSN'][index]] if val != 'NM' ][0]

In [251]:
corresponding_payment_of_min_values = {}
for ssn in unique_SSN:
    corresponding_payment_of_min_values[ssn] = list(df_payment_of_min_amount[df_payment_of_min_amount['SSN'] == ssn]['Payment_of_Min_Amount'].unique())

In [252]:
dict(itertools.islice(corresponding_payment_of_min_values.items(),10))

{821000265: ['No'],
 4075839: ['No'],
 486853974: ['No'],
 72316145: ['No'],
 615067821: ['Yes'],
 612708987: ['No'],
 411510676: ['No'],
 500926408: ['Yes'],
 70191622: ['No'],
 366681681: ['No']}

In [253]:
df_payment_of_min_amount['Payment_of_Min_Amount'].value_counts()

Yes    59432
No     40568
Name: Payment_of_Min_Amount, dtype: int64

### PreProcessing Total_EMI_per_month

In [254]:
df_total_emi_per_month = df_payment_of_min_amount.copy()

In [255]:
df_total_emi_per_month['Total_EMI_per_month'].nunique()

14950

In [256]:
df_total_emi_per_month['Total_EMI_per_month'].isna().sum()

0

### PreProcessing Amount_invested_monthly

In [66]:
df_amount_invested_monthly = df_total_emi_per_month.copy()

In [67]:
df_amount_invested_monthly["Amount_invested_monthly"].nunique()

91049

### PreProcessing Payment_Behaviour

In [69]:
df_payment_behaviour = df_amount_invested_monthly.copy()

In [70]:
df_payment_behaviour["Payment_Behaviour"].nunique()

7

### PreProcessing Monthly_Balance

In [71]:
df_monthly_balance = df_payment_behaviour.copy()

In [72]:
df_payment_behaviour["Payment_Behaviour"].nunique()

7

### PreProcessing Credit_Score

In [73]:
df_credit_score = df_payment_behaviour.copy()

In [74]:
df_credit_score["Credit_Score"].nunique()

3